src/librustc_mir/lib.rs

```diff
@@ -51,19 +51,22 @@ mod shim;
 pub mod transform;
 pub mod util;
 pub mod interpret;
-pub mod monomorphize;
 pub mod const_eval;
+pub mod monomorphize;
 
-use rustc::ty::query::Providers;
+use rustc::ty::{ParamEnvAnd, query::Providers};
 
 pub fn provide(providers: &mut Providers<'_>) {
     borrow_check::provide(providers);
     shim::provide(providers);
+    monomorphize::provide(providers);
     transform::provide(providers);
-    monomorphize::partitioning::provide(providers);
     providers.const_eval = const_eval::const_eval_provider;
     providers.const_eval_raw = const_eval::const_eval_raw_provider;
     providers.check_match = hair::pattern::check_match;
+    providers.const_field = |tcx, ParamEnvAnd { param_env, value: (field, value) }| {
```

Comment:
Do we do this anywhere else? I could've sworn it looked different but now I'm not sure where to tell you to look for an example (`rustc::ty::layout` maybe?)

Reply:
It looks like it's done in a lot of places, but usually the key isn't destructured. I can move the destructuring to a `let` binding in the body if you like.

```console
$ rg --iglob '*.rs' 'providers.*\|tcx'
src/librustc_codegen_llvm/attributes.rs
317:    providers.target_features_whitelist = |tcx, cnum| {
337:    providers.wasm_import_module_map = |tcx, cnum| {

src/librustc_codegen_ssa/base.rs
860:    providers.backend_optimization_level = |tcx, cratenum| {
893:    providers.dllimport_foreign_items = |tcx, krate| {
918:    providers.is_dllimport_foreign_item = |tcx, def_id| {

src/librustc_mir/monomorphize/partitioning.rs
969:    providers.is_codegened_item = |tcx, def_id| {
975:    providers.codegen_unit = |tcx, name| {

src/librustc/ty/context.rs
3048:    providers.in_scope_traits_map = |tcx, id| tcx.gcx.trait_map.get(&id);
3049:    providers.module_exports = |tcx, id| tcx.gcx.export_map.get(&id).map(|v| &v[..]);
3050:    providers.crate_name = |tcx, id| {
3054:    providers.get_lib_features = |tcx, id| {
3058:    providers.get_lang_items = |tcx, id| {
3062:    providers.upvars = |tcx, id| tcx.gcx.upvars.get(&id).map(|v| &v[..]);
3063:    providers.maybe_unused_trait_import = |tcx, id| {
3066:    providers.maybe_unused_extern_crates = |tcx, cnum| {
3070:    providers.names_imported_by_glob_use = |tcx, id| {
3075:    providers.stability_index = |tcx, cnum| {
3079:    providers.lookup_stability = |tcx, id| {
3084:    providers.lookup_deprecation_entry = |tcx, id| {
3089:    providers.extern_mod_stmt_cnum = |tcx, id| {
3093:    providers.all_crate_nums = |tcx, cnum| {
3097:    providers.postorder_cnums = |tcx, cnum| {
3101:    providers.output_filenames = |tcx, cnum| {
3105:    providers.features_query = |tcx, cnum| {
3109:    providers.is_panic_runtime = |tcx, cnum| {
3113:    providers.is_compiler_builtins = |tcx, cnum| {

src/test/run-make-fulldeps/hotplug_codegen_backend/the_backend.rs
50:        providers.target_features_whitelist = |tcx, _cnum| {

src/librustc_typeck/check/method/suggest.rs
844:    providers.all_traits = |tcx, cnum| {
```

Reply:
Sorry, I was referring to the `ParamEnvAnd` destructuring, since I could've sworn I've never seen it used.

Reply:
Yeah, this is how we actually do this: https://github.com/rust-lang/rust/blob/fe5f42cdb88d8ce31f746130099321e7c95e1ef0/src/librustc/ty/util.rs#L940-L944